presto-iceberg/src/main/java/com/facebook/presto/iceberg/statistics/KllHistogram.java

```diff
@@ -0,0 +1,179 @@
+import org.apache.datasketches.common.ArrayOfItemsSerDe;
+import org.apache.datasketches.common.ArrayOfLongsSerDe;
+import org.apache.datasketches.common.ArrayOfStringsSerDe;
+import org.apache.datasketches.kll.KllItemsSketch;
+import org.apache.datasketches.memory.Memory;
+
+import java.util.Comparator;
+import java.util.function.Function;
+
+import static com.facebook.presto.common.type.DateType.DATE;
+import static com.facebook.presto.common.type.DoubleType.DOUBLE;
+import static com.facebook.presto.common.type.TypeUtils.isExactNumericType;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_ARGUMENTS;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Verify.verify;
+import static java.nio.ByteOrder.LITTLE_ENDIAN;
+import static java.util.Objects.requireNonNull;
+import static org.apache.datasketches.quantilescommon.QuantileSearchCriteria.EXCLUSIVE;
+import static org.apache.datasketches.quantilescommon.QuantileSearchCriteria.INCLUSIVE;
+
+public class KllHistogram
+        implements ConnectorHistogram
+{
+    // since the actual type parameter is only known at runtime, we can't concretely specify it
+    private final KllItemsSketch<Object> sketch;
+    private final Type type;
+    private final Function<Object, Double> toDouble;
+    private final Function<Double, Object> fromDouble;
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    @JsonCreator
+    public KllHistogram(@JsonProperty("sketch") Slice bytes, @JsonProperty("type") Type type)
+    {
+        verify(isKllHistogramSupportedType(type), "histograms do not currently support type " + type.getDisplayName());
+        this.type = requireNonNull(type, "type is null");
+        SketchParameters parameters = getSketchParameters(type);
+        // the actual sketch can only accept the same object types which generated it
+        // however, the API can only accept or generate double types. We cast the inputs
+        // and results to/from double to satisfy the underlying sketch type.
+        if (type.getJavaType().isAssignableFrom(double.class)) {
+            toDouble = x -> (double) x;
+            fromDouble = x -> x;
+        }
+        else if (type.getJavaType().isAssignableFrom(long.class)) {
+            // dual cast to auto-box/unbox from Double/Long for sketch
+            toDouble = x -> (double) (long) x;
+            fromDouble = x -> (long) (double) x;
+        }
+        else {
+            throw new PrestoException(INVALID_ARGUMENTS, "can't create kll sketch from type: " + type);
+        }
+        sketch = KllItemsSketch.wrap(Memory.wrap(bytes.toByteBuffer(), LITTLE_ENDIAN), parameters.getComparator(), parameters.getSerde());
+    }
+
+    public static boolean isKllHistogramSupportedType(Type type)
+    {
+        return isExactNumericType(type) ||
+                type.equals(DOUBLE) ||
+                type.equals(DATE) ||
+                type instanceof ShortDecimalType;
+    }
```

Comment:
I notice your comment in `IcebergDistributedTestBase` for explaining the support of types. As I understand your description, in theory we can support types of `real`/`time`/`timestamp` in future, but that need a little more works. Is it right?

Reply:
Yes I am working on this in: https://github.com/prestodb/presto/pull/23114